neonvm/controllers/virtualmachine_controller.go

```diff
@@ -201,11 +205,47 @@ func (r *VirtualMachineReconciler) doFinalizerOperationsForVirtualMachine(ctx co
 	// to set the ownerRef which means that the Deployment will be deleted by the Kubernetes API.
 	// More info: https://kubernetes.io/docs/tasks/administer-cluster/use-cascading-deletion/
 
+	log := log.FromContext(ctx)
+
 	// The following implementation will raise an event
 	r.Recorder.Event(virtualmachine, "Warning", "Deleting",
 		fmt.Sprintf("Custom Resource %s is being deleted from the namespace %s",
 			virtualmachine.Name,
 			virtualmachine.Namespace))
+
+	// Release overlay IP address
+	if virtualmachine.Spec.ExtraNetwork != nil {
+		// Create IPAM object
+		nadName, err := nadIpamName()
+		if err != nil {
+			// ignore error
+			log.Info("ignored error", err)
+			return nil
+		}
+		nadNamespace, err := nadIpamNamespace()
+		if err != nil {
+			// ignore error
+			log.Info("ignored error", err)
+			return nil
+		}
+		ipam, err := ipam.New(ctx, nadName, nadNamespace)
+		if err != nil {
+			// ignore error
+			log.Info("ignored error", err)
+			return nil
+		}
+		defer ipam.Close()
+		ip, err := ipam.ReleaseIP(ctx, virtualmachine.Name, virtualmachine.Namespace)
+		if err != nil {
+			// ignore error
+			log.Info("fail to release IP", "error", err)
+			return nil
+		}
```

Comment:
structured logs already have controller type and vm (vm migration) object name. so not needed add it additionally. I will fix log.Info format, thanks.

Regarding `ignored error` - initially I didn't want print any errors. It mostly for potential debug. Not sure what better - leave these logs or remove at all. Now it will be `log.Error`

Reply:
about structured logs, example:

in the code: `log.Info("Removing Finalizer from Migration")`

in the logs:
```json
{"level":"info","ts":"2023-06-08T11:06:35.101Z","msg":"Removing Finalizer from Migration","controller":"virtualmachinemigration","controllerGroup":"vm.neon.tech","controllerKind":"VirtualMachineMigration","VirtualMachineMigration":{"name":"vm-stress-0099-035","namespace":"default"},"namespace":"default","name":"vm-stress-0099-035","reconcileID":"30cc7928-7bed-4a4c-b22d-9a8d82c5f414"}
```